test/src/specs/mining/uncle.rs

```diff
@@ -0,0 +1,196 @@
+            .set_uncles(vec![uncle.as_uncle()])
+            .build();
+        target_node.submit_block(&block.data());
+    }
+}
+
+pub struct UncleInheritFromForkUncle;
+
+impl Spec for UncleInheritFromForkUncle {
+    crate::name!("uncle_inherit_from_fork_uncle");
+
+    crate::setup!(num_nodes: 2, connect_all: false);
+
+    // Case: A uncle inherited from a fork-uncle in side fork is invalid, because that breaks
+    //       the uncle rule "B1's parent is either B2's ancestor or embedded in B2 or its ancestors
+    //       as an uncle"
+    fn run(&self, net: Net) {
+        let target_node = &net.nodes[0];
+        let feed_node = &net.nodes[1];
+        net.exit_ibd_mode();
+
+        info!("(1) Build a chain which embedded `uncle_parent` as an uncle");
+        let uncle_parent = construct_uncle(target_node);
+        target_node.submit_block(&uncle_parent.data());
+
+        let uncle_child = uncle_parent
+            .as_advanced_builder()
+            .number((uncle_parent.number() + 1).pack())
+            .parent_hash(uncle_parent.hash())
+            .timestamp((uncle_parent.timestamp() + 1).pack())
+            .epoch(
+                {
+                    let parent_epoch = uncle_parent.epoch();
+                    let epoch_number = parent_epoch.number();
+                    let epoch_index = parent_epoch.index();
+                    let epoch_length = parent_epoch.length();
+                    EpochNumberWithFraction::new(epoch_number, epoch_index + 1, epoch_length)
+                }
+                .pack(),
+            )
+            .build();
+
+        info!("(2) Force reorg, so that `uncle_parent` become a fork-uncle");
+        let longer_fork = (0..=target_node.get_tip_block_number()).map(|_| {
+            let block = feed_node.new_block(None, None, None);
+            feed_node.submit_block(&block.data());
+            block
+        });
+        longer_fork.for_each(|block| {
+            target_node.submit_block(&block.data());
+        });
+
+        info!("(3) Submit block with `uncle`, which is inherited from fork-uncle `uncle_parent`, should be failed");
+        let block = target_node
+            .new_block_builder(None, None, None)
+            .set_uncles(vec![uncle_child.as_uncle()])
+            .build();
+        let ret = target_node
+            .rpc_client()
+            .submit_block("0".to_owned(), block.data().into());
+        assert!(ret.is_err(), "{:?}", ret);
+        let err = ret.unwrap_err();
+        assert!(err.to_string().contains("DescendantLimit"), "{:?}", err);
+
+        info!("(4) Add all the fork-blocks as uncle blocks into the chain. And now re-submit block with `uncle_child` should be success");
+        until_no_uncles_left(target_node);
+        let block = target_node
+            .new_block_builder(None, None, None)
+            .set_uncles(vec![uncle_child.as_uncle()])
+            .build();
+        target_node.submit_block(&block.data());
+    }
+}
```

Comment:
It's not easy to get all the things of the above two test specs, what about adding an ASCII graph.

Reply:
Many of integration-tests try to simulate complicated scenarios.
Yes, I will consider adding more descriptions to complicated cases. And ASCII graph is a good idea.